catalog/core/catalog-core-subscriptionstore/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionCache.java

```diff
@@ -0,0 +1,234 @@
+
+        if (!type.equals(metadata.getType())) {
+            return null;
+        }
+
+        return metadata.getSubscription()
+                .orElseThrow(() -> new SubscriptionStoreException(format(
+                        "Subscription [%s] was not cached",
+                        metadata.getId())));
+    }
+
+    @Override
+    public String insert(Subscription subscription, String type, String messageBody,
+            String callbackUrl) {
+        notNull(subscription, "subscription cannot be null");
+        notEmpty(type, "type cannot be null or empty");
+        notEmpty(messageBody, "messageBody cannot be null or empty");
+        notEmpty(callbackUrl, "callbackUrl cannot be null or empty");
+
+        SubscriptionMetadata metadata = new SubscriptionMetadata(type, messageBody, callbackUrl);
+
+        createSubscriptionLocally(metadata, subscription);
+        subscriptionStore.insert(metadata);
+
+        return metadata.getId();
+    }
+
+    @Override
+    public void update(Subscription subscription, String type, String messageBody,
+            String callbackUrl, String subscriptionId) {
+        notNull(subscription, "subscription cannot be null");
+        notEmpty(type, "type cannot be null or empty");
+        notEmpty(messageBody, "messageBody cannot be null or empty");
+        notEmpty(callbackUrl, "callbackUrl cannot be null or empty");
+        notEmpty(subscriptionId, "subscriptionId cannot be null or empty");
+
+        if (!contains(subscriptionId, type)) {
+            LOGGER.debug("Target for subscription update [{}|{}] does not exist",
+                    subscriptionId,
+                    type);
+            return;
+        }
+
+        delete(subscriptionId, type);
+
+        SubscriptionMetadata metadata = new SubscriptionMetadata(type,
+                messageBody,
+                callbackUrl,
+                subscriptionId);
+
+        createSubscriptionLocally(metadata, subscription);
+        subscriptionStore.insert(metadata);
+    }
+
+    @Nullable
+    @Override
+    public Subscription delete(String subscriptionId, String type) {
+        notEmpty(subscriptionId, "subscriptionId cannot be null or empty");
+        notEmpty(type, "type cannot be null or empty");
+
+        if (!contains(subscriptionId, type)) {
+            return null;
+        }
+
+        SubscriptionMetadata metadata = deleteSubscriptionLocally(subscriptionId);
+        subscriptionStore.delete(subscriptionId);
+
+        return metadata.getSubscription()
+                .orElseThrow(() -> new SubscriptionStoreException(format(
+                        "Subscription [%s] was not cached",
+                        metadata.getId())));
+    }
+
+    @Override
+    public boolean contains(String subscriptionId, String type) {
+        return subscriptions.containsKey(subscriptionId) && type.equals(subscriptions.get(
+                subscriptionId)
+                .getType());
+    }
```

Comment:
Take note: We need to handle the case where the clever client passes in a subscription ID that exists but doesn't belong to them.